drivers/misc/ipts/ipts-companion.c

```diff
@@ -0,0 +1,203 @@
+#include <linux/firmware.h>
+#include <linux/ipts.h>
+#include <linux/ipts-binary.h>
+#include <linux/ipts-companion.h>
+#include <linux/mutex.h>
+
+#include "ipts.h"
+#include "ipts-companion.h"
+#include "ipts-params.h"
+
+#define IPTS_FW_PATH_FMT "intel/ipts/%s"
+#define IPTS_FW_CONFIG_FILE "ipts_fw_config.bin"
+
+ipts_companion_t *ipts_companion;
+DEFINE_MUTEX(ipts_companion_lock);
+
+bool ipts_companion_available(void)
+{
+	bool ret;
+	mutex_lock(&ipts_companion_lock);
```

Comment:
is there a reason you picked mutex over spin here?

Reply:
Because qzed told me so when writing the original implementation :P (in `ipts-fw.c` which was deleted).

I am not that familiar with synchronisation inside of the kernel, so I just kept it like that.

Reply:
My reasoning was mostly: Mutex is the standard tool and it's initialization code, so we don't need the performance a spin-lock would give us here. Also the lock is being held during IO operations (`request_firmware`), so not really small critical sections.